### PR TITLE
[SDEV-1692] Fix and refactor registrar classes

### PR DESCRIPTION
**Pull request by tepals**
[fix] ShiftRegistrar._compute_registration a position in pixels was compared to a position in meters
    [refactor] Renaming of registered_positions to registered_positions_m and addition of registered_positions_px
    [refactor] Let GlobalShiftRegistrar inherit from the ShiftRegistrar class.
    [fix] shift registrar synthetic test case

### CODE REVIEW

**Review by patrickcleeve2 — APPROVED** [review]

File: src/odemis/acq/stitching/_registrar.py [hunk 1]
```diff
@@ -96,16 +99,16 @@ def __init__(self):
...
-        self.acqOrder = []
+        self.acq_order = []
 
         # Shift between main tile and dependent tiles
         self.shift_tile_dep_tiles = []  # N x K list. N: number of tiles, K: number of dep_tiles.
 
         self.px_size = None  # Tuple of 2 ints. (X,Y) pixel size of the current tile in m/px.
         self.size = None  # Tuple of 2 ints. (Y,X) shape of the current tile in px.
-        self.posX = None  # int. Position of the current tile in the grid. (0,0) is the top left, posX
+        self.pos_x = None  # int. Position of the current tile in the grid. (0,0) is the top left, posX
```
Comment (nandishjpatel) [inline src/odemis/acq/stitching/_registrar.py:109]:
```suggestion
        self.pos_x = None  # int. Position of the current tile in the grid. (0,0) is the top left, pos_x
```

File: src/odemis/acq/stitching/_registrar.py [hunk 2]
```diff
@@ -124,100 +127,105 @@ def addTile(self, tile, dependent_tiles=None):
...
+        return self.registered_positions_m, dep_tile_positions
+
+    def _insert_tile_to_grid(self, tile):
+        """
+        Stores the tile at the proper place in the grid. If necessary, the grid is
+        extended to accommodate the new tile.
+
+        :param tile: (DataArray) tile to be inserted
+        :returns: (int, int) row, col grid position of the tile
+        :updates self.tiles, self.posX, self.posY, self.ovrlp, self.px_size, self.size:
```
Comment (nandishjpatel) [inline src/odemis/acq/stitching/_registrar.py:171]:
```suggestion
        :updates self.tiles, self.pos_x, self.pos_y, self.ovrlp, self.px_size, self.size:
```

File: src/odemis/acq/stitching/_registrar.py [hunk 3]
```diff
@@ -614,52 +603,57 @@ def _get_shift(self, prev_tile, tile):
...
+            logging.info(f"Calculated shift {meas_tile_dist_px} is larger than the overlap size {overlap}, "
                          f"using expected position instead.")
-            return exp_shift, 0
+            return exp_tile_dist_px, 0
 
-        return shift_total, ncc
+        return shift_px, ncc
 
-    def _compute_registration(self, row, col):
+    def _compute_registration(self, tile, row, col):
```
Comment (nandishjpatel) [inline src/odemis/acq/stitching/_registrar.py:656]: Is this param used?

File: src/odemis/acq/stitching/_registrar.py [hunk 4]
```diff
@@ -124,100 +127,105 @@ def addTile(self, tile, dependent_tiles=None):
...
+
+        # Insert new tile either to the right or to the bottom of the
+        # closest tile.
+        ver_diff = md_pos[1] - md_pos_prev[1]
+        hor_diff = md_pos[0] - md_pos_prev[0]
+        if abs(ver_diff) > abs(hor_diff):
+            if ver_diff < 0:
+                self.pos_y = pos_prev[0] + 1
+                self.pos_x = pos_prev[1]
+                if len(self.registered_positions_px) <= self.pos_y:
```
Comment (nandishjpatel) [inline src/odemis/acq/stitching/_registrar.py:201]:
```suggestion
                if len(self.registered_positions_px[1]) <= self.pos_y:
```

File: src/odemis/acq/stitching/_registrar.py [hunk 5]
```diff
@@ -124,100 +127,105 @@ def addTile(self, tile, dependent_tiles=None):
...
+                    self._updateGrid("y")  # extend grid in y direction
+            else:
+                self.pos_y = pos_prev[0] - 1
+                self.pos_x = pos_prev[1]
+        else:
+            if hor_diff > 0:
+                self.pos_y = pos_prev[0]
+                self.pos_x = pos_prev[1] + 1
+                if len(self.registered_positions_px[0]) <= self.pos_x:
```
Comment (nandishjpatel) [inline src/odemis/acq/stitching/_registrar.py:210]: This check is a bit different from line 201?

File: src/odemis/acq/stitching/_registrar.py [hunk 6]
```diff
@@ -87,7 +89,8 @@ def __init__(self):
         # initialize grid to 1x1. The size will increase as new tiles are
         # added.
         self.nx = 1  # (int) total number of cols
-        self.registered_positions = [[None]]  # (None or tuple of 2 floats) list of calculated positions
+        self.registered_positions_px = [[None]]  # (None or tuple of 2 floats) list of calculated shifts in px
+        self.registered_positions_m = [[None]]  # (None or tuple of 2 floats) list of calculated positions in m
```
Comment (nandishjpatel) [inline src/odemis/acq/stitching/_registrar.py:93]: Do you actually need this? Is this for better understanding for the user? I think returning `tile_positions` will just do fine in line 162

**Review by nandishjpatel — APPROVED** [review]

File: src/odemis/acq/stitching/_registrar.py (hunk 6 above)
Comment (pieleric) [reply src/odemis/acq/stitching/_registrar.py:93]: Indeed, this doesn't seem used any where. Also, the comment seem to imply the difference between the two is just a coefficient factor, but one is relative to the center of the first tile, while the other one is in absolute coordinates. Maybe worth mentioning?

File: src/odemis/acq/stitching/_registrar.py [hunk 7]
```diff
@@ -284,9 +291,6 @@ def _estimateMatch(self, imageA, imageB, shift):
         Returns an estimation of the similarity between the given images
         when the second is shifted by the shift value. It is used to assess 
         the quality of a shift measurement by giving the shifted image.
-        imageA:
-        imageB:
-        shift: shift after phase correlation
```
Comment (pieleric) [inline src/odemis/acq/stitching/_registrar.py]: Why removing this doc? I'd suggest to at least document their types.

File: src/odemis/acq/stitching/_registrar.py (hunk 3 above)
Comment (pieleric) [reply src/odemis/acq/stitching/_registrar.py:656]: I think the idea is that this method has the same signature as the parent class.

**Review by pieleric — CHANGES_REQUESTED** [review]

File: src/odemis/acq/stitching/_registrar.py (hunk 3 above)
Comment (tepals) [reply src/odemis/acq/stitching/_registrar.py:656]: that's indeed the idea. But this also allows me to remove line 666, which I will do.

File: src/odemis/acq/stitching/_registrar.py (hunk 5 above)
Comment (tepals) [reply src/odemis/acq/stitching/_registrar.py:210]: That is correct. It is a list of lists, therefore `len(self.registered_positions_px)` gives the amount of vertical tiles, and `len(self.registered_positions_px[0])` gives the number of horizontal tiles in the first line

**Review by nandishjpatel — APPROVED** [review]

File: src/odemis/acq/stitching/test/registrar_test.py [hunk 8]
```diff
@@ -607,46 +612,61 @@ class TestGlobalShiftRegistrar(unittest.TestCase):
...
-                    start = numpy.add((int(0.2 * size[0]), 0), shift)
-                    end = (start[0] + tsize[0], start[1] + tsize[1])
-                    tile2 = img[start[0]:end[0], start[1]:end[1]]
-
-                    px_size = (1e-6, 2e-5)
+                    # create a black image
+                    img = numpy.zeros(img_size, dtype=dtype)
+                    for _ in range(int(img_size[0] / 3)):
+                        # Randomly draw vertical and horizontal lines
+                        is_vertical = numpy.random.choice([True, False])
```
Comment (pieleric) [inline src/odemis/acq/stitching/test/registrar_test.py:633]: No need to use numpy for that. `random.choice()` works as fine.

File: src/odemis/acq/stitching/test/registrar_test.py [hunk 9]
```diff
@@ -607,46 +612,61 @@ class TestGlobalShiftRegistrar(unittest.TestCase):
...
+                        if is_vertical:
+                            # Draw a vertical line
+                            x = numpy.random.randint(0, img_size[0])
+                            color = numpy.random.randint(1, 256)
```
Comment (pieleric) [inline src/odemis/acq/stitching/test/registrar_test.py:637]: `random.randint() should as fine here.

File: src/odemis/acq/stitching/test/registrar_test.py [hunk 10]
```diff
@@ -607,46 +612,61 @@ class TestGlobalShiftRegistrar(unittest.TestCase):
...
+                            img[x:x + 2, :] = color
+                        else:
+                            # Draw a horizontal line
+                            y = numpy.random.randint(0, img_size[1])
+                            color = numpy.random.randint(1, 256)
```
Comment (pieleric) [inline src/odemis/acq/stitching/test/registrar_test.py:642]: just use `random.randint()` here too.

File: src/odemis/acq/stitching/test/registrar_test.py [hunk 11]
```diff
@@ -311,46 +301,61 @@ class TestShiftRegistrar(unittest.TestCase):
 
     def setUp(self):
         random.seed(1)
+        numpy.random.seed(10)
```
Comment (pieleric) [inline src/odemis/acq/stitching/test/registrar_test.py:304]: I think you don't need anything specific from numpy.random. In this case, you could drop this line (and its clone in the other test case).

**Review by pieleric — CHANGES_REQUESTED** [review]

File: src/odemis/acq/stitching/_registrar.py [hunk 12]
```diff
@@ -124,100 +127,105 @@ def addTile(self, tile, dependent_tiles=None):
...
+                                   first_position_m[1] - shift[1] * self.px_size[1]))
 
         # Return positions for dependent tiles
         for t, sdts in zip(tile_positions, self.shift_tile_dep_tiles):
             dts = []  # dependent tiles for tile
             for sdt in sdts:
                 dts.append((t[0] + sdt[0], t[1] + sdt[1]))
             dep_tile_positions.append(dts)
+        self.registered_positions_m = tile_positions
+        return self.registered_positions_m, dep_tile_positions
```
Comment (tepals) [inline src/odemis/acq/stitching/_registrar.py:162]: this should still be removed

File: src/odemis/acq/stitching/test/registrar_test.py (hunk 11 above)
Comment (tepals) [reply src/odemis/acq/stitching/test/registrar_test.py]: thanks, good catch.

**Pull request merged**